### PR TITLE
Fetch underlay configuration from the backend

### PR DESCRIPTION
**Pull request by tjennison-google**
It is stored for later access to hints and other information.

### CODE REVIEW

File: ui/src/app.tsx [hunk 1]
```diff
@@ -17,25 +22,65 @@ import Overview from "./overview";
...
+            }
+            return entitiesApi.listEntities({ underlayName: u.name });
+          })
+        ).then((entitiesResList) => {
+          dispatch(
+            setUnderlays(
+              entitiesResList.map((entitiesRes, i) => {
+                const name = res.underlays?.[i]?.name;
+                if (!name) {
+                  throw "Unnamed underlay.";
```
Comment (calbach) [inline ui/src/app.tsx:53]: nit: this should be unreachable due to above check at L42

File: ui/src/app.tsx [hunk 2]
```diff
@@ -17,25 +22,65 @@ import Overview from "./overview";
...
+                }
+                if (!entitiesRes.entities) {
```
Comment (calbach) [inline ui/src/app.tsx:55]: Should this check length as well?

File: ui/src/app.tsx [hunk 3]
```diff
@@ -17,25 +22,65 @@ import Overview from "./overview";
...
+                  throw `No entities in underlay ${name}`;
+                }
+
+                return {
+                  name: name,
```
Comment (calbach) [inline ui/src/app.tsx:60]: nit: `name,`

File: ui/.env.test [hunk 4]
```diff
@@ -0,0 +1 @@
+REACT_APP_USE_FAKE_API=y
```
[the remark this answers is not in the document]
Comment (tjennison-google) [reply ui/.env.test:1]: The flag sets up the fake API and `.env.test` is pulled in whenever tests are being run. Previously none of the tests resulted in network requests so it wasn't necessary before.

File: ui/src/app.tsx [hunk 5]
```diff
@@ -17,25 +22,65 @@ import Overview from "./overview";
...
 export default function App(props: AppProps) {
+  const dispatch = useAppDispatch();
+  const underlaysApi = useContext(UnderlaysApiContext);
+  const entitiesApi = useContext(EntitiesApiContext);
+
+  const underlaysState = useAsyncWithApi(
+    useCallback(() => {
+      return underlaysApi.listUnderlays({}).then((res) => {
+        if (!res?.underlays || res.underlays.length == 0) {
+          throw "No underlays are configured.";
```
[the remark this answers is not in the document]
Comment (tjennison-google) [reply ui/src/app.tsx]: No, if throwing Error is standard that's what I'll do as well.

File: ui/src/app.tsx (hunk 1 above)
Comment (tjennison-google) [reply ui/src/app.tsx]: Yes, but unfortunately the compiler can't confirm it so it complains without the extra check.

File: ui/src/app.tsx (hunk 2 above)
Comment (tjennison-google) [reply ui/src/app.tsx]: I don't think it needs to. For our current schemas we'd always expect there to be entities but it doesn't seem like an error in general.

File: ui/src/app.tsx (hunk 3 above)
Comment (tjennison-google) [reply ui/src/app.tsx]: Done.

File: ui/src/app.tsx [hunk 6]
```diff
@@ -17,25 +22,65 @@ import Overview from "./overview";
...
   entityName: string;
 };
 
 export default function App(props: AppProps) {
+  const dispatch = useAppDispatch();
+  const underlaysApi = useContext(UnderlaysApiContext);
+  const entitiesApi = useContext(EntitiesApiContext);
+
+  const underlaysState = useAsyncWithApi(
+    useCallback(() => {
```
[the remark this answers is not in the document]
Comment (tjennison-google) [reply ui/src/app.tsx]: Done. I agree it's more readable.

**Review by calbach — APPROVED** [review]

**Pull request merged**